pkg/controllers/dashboard/systemcharts/controller.go

```diff
@@ -232,14 +240,47 @@ func (h *handler) getChartsToInstall() []*chart.Definition {
 // when a specific event occurs on the target deployment. It is currently used to manage
 // the migration from the legacy k3s-based-upgrader app to the system-upgrade-controller app.
 func (h *handler) onDeployment(_ string, d *k8sappsv1.Deployment) (*k8sappsv1.Deployment, error) {
+	var found bool
 	if d != nil && d.Namespace == namespace.System && d.Name == sucDeploymentName {
+		appName, ok := d.Labels[legacyAppLabel]
+		if ok {
+			if appName == k3sbasedupgrade.K3sAppName || appName == k3sbasedupgrade.Rke2AppName {
+				found = true
+			}
+		}
+	}
+	if found {
```

Comment:
Suggesting to go with "return early" strategy for better readability here. Basically if deployment doesn't need to be updated, exit out of this func. This will reduce nesting in the code below and will make it a bit readable (reviewing this PR I initially thought `else` part on line 275 will be executed for ALL deployments but then it's within the `if found` block actually).